Use storage.local instead of localStorage

danfinlay opened the pull request:
Fixes #3076, needs thorough QA:

- Create a vault on `master`, rebuild on this branch, see if MM continues seamlessly.
- Try destroying the old `localStorage` data manually, see if MM continues seamlessly.
- Do [full QA](https://consensys.quip.com/YmJuANWLkEAv) on this branch, to ensure it's internally reliable.
- Maybe try filling up storage manually somehow. Could be by creating tons of accounts or tons of testnet transactions that publish large contracts.

heyellieday committed:
Add ExtensionStore and add basic store instance syncing to main controller

heyellieday committed:
handle situation where storage.sync is disabled in certain versions of firefox

heyellieday committed:
conditionally use extension store if supported or enabled

danfinlay committed:
Add unlimitedStorage option to manifest

danfinlay committed:
Modify @heyellieday's work to use storage.local to replace main storage

danfinlay committed:
Merge branch 'UseStorageLocal' into i3076-UseStorageLocalInstead

danjm committed:
Handle errors when getting and setting to localStore.

danjm committed:
Check that extension.storage exists before attempting to call methods on it.

danjm committed:
Finish tests for extension-store fetch and sync.

danfinlay committed:
Merge pull request #3079 from MetaMask/i3076-UseStorageLocalInstead-danjm-1

Additions to the UseStorageLocalInstead branch

danfinlay committed:
Bump changelog

kumavis commented on app/scripts/background.js:
```diff
@@ -100,6 +117,17 @@ function setupController (initState) {
+      try {
+        localStore.set(state)
+      } catch (err) {
+        log.error('error setting state in local store:', err)
```
i dont think these get submitted to sentry (but maybe they do)

danfinlay replied:
What does get submitted for sure?

danfinlay replied:
(we should document our correct logging practice)

kumavis replied:
im not certain atm, but you can test it with the `Test-Metamask` sentry project on dev

kumavis commented on app/scripts/lib/extension-store.js:
```diff
@@ -0,0 +1,39 @@
+  }
+  async fetch() {
+    return new Promise((resolve) => {
+      extension.storage && extension.storage.sync.get(KEYS_TO_SYNC, (data) => {
```
we wanted to drop `KEYS_TO_SYNC`, right?

kumavis commented on app/scripts/background.js:
```diff
@@ -52,6 +54,20 @@ async function loadStateFromPersistence () {
+      log.error('error fetching state from local store:', err)
+    }
+
+   // TODO: handle possible exceptions (https://developer.chrome.com/apps/runtime#property-lastError)
```
looks like this is somewhat handled in `ExtensionStore`
```js
 const lastError = extension.runtime.lastError
```

kumavis replied:
I'm hesitant to merge a TODO in here, especially around error handling

kumavis commented on app/scripts/lib/local-store.js:
```diff
@@ -0,0 +1,25 @@
+      log.error('Storage local API not available.')
+    }
+  }
+  async get() {
```
doesnt need to be marked as `async` when manually returning a promise but maybe this is a good practice anyways

kumavis reviewed (commented):
Nice -- looking forward to it

added some questions / comments

danfinlay commented on app/scripts/lib/extension-store.js:
```diff
@@ -0,0 +1,39 @@
+const KEYS_TO_SYNC = ['KeyringController', 'PreferencesController']
+const FIREFOX_SYNC_DISABLED_MESSAGE = 'Please set webextensions.storage.sync.enabled to true in about:config'
+
+const handleDisabledSyncAndResolve = (resolve, toResolve) => {
```
This is part of the original Ellieday PR, we aren't actually using this file in this PR, so I'm not sure I want to bother cleaning it up, I'd rather either:
- Delete the file and its test
- Ignore it for now

tmashuang commented on app/manifest.json:
```diff
@@ -60,6 +60,7 @@
     "http://localhost:8545/",
     "https://*.infura.io/"
     ],
+  "unlimitedStorage": true,
```
`unlimitedStorage` should be added to the permissions, the same as `storage`. Referenced: [Declare Permissions](https://developer.chrome.com/extensions/declare_permissions)

danfinlay committed:
Corrected unlimitedStorage permission

danfinlay committed:
Remove extension-store since we aren't using it yet

kumavis commented:
@heyellieday we used some of your code in this PR so we'd like to credit you with partial bounty reward of the original bounty (synced storage). Can you give an ethereum address here or emailed to `aaron at kumavis dot me` thanks/

danjm committed:
Remove already handled TODO comment.

kumavis commented:
cool, just needs QA then

danfinlay committed:
Remove redundant async modifiers

danfinlay committed:
Fix retrieval of object

danfinlay committed:
Remove local storage writes, add log

danfinlay committed:
Debounce storage to avoid crashing pump

danfinlay commented:
Added to that branch:
- No longer writes to `localStorage`.
- Fixed issue where quantity of write attempts overwhelmed our `pump` to `storage.local` stream.

danfinlay committed:
Increase storage debounce value

heyellieday commented:
@kumavis, sorry, I missed these mentions!

I'm glad some of my work was able to be used for this issue and thanks for crediting me part of the bounty! Here's an ethereum address to send to: 0x1ab4C79D96a05b64742AE1B0213B84fB984917C8.

I'm hoping I can help out with more in the future :)

danfinlay committed:
Merge branch 'master' into i3076-UseStorageLocalInstead

danfinlay committed:
Merge branch 'master' into i3076-UseStorageLocalInstead

danfinlay committed:
Increase storage debounce to 2s

danfinlay committed:
Update yarn.lock

danfinlay committed:
Some fixes for new storage strategy

- Don't persist undefined data
- Write to new storage strategy without waiting for completion.
- Continue writing to localStorage as fallback.

kumavis commented on app/scripts/background.js:
```diff
@@ -65,11 +69,34 @@ async function loadStateFromPersistence () {
+    }
+
+    if (Object.keys(localData).length > 0) {
+      versionedData = localData
```
why do we do this
please add a comment

danfinlay replied:
Comment added.

kumavis commented on app/scripts/background.js:
```diff
@@ -65,11 +69,34 @@ async function loadStateFromPersistence () {
   const migrator = new Migrator({ migrations })
   // read from disk
-  let versionedData = diskStore.getState() || migrator.generateInitialState(firstTimeState)
+  versionedData = diskStore.getState() || migrator.generateInitialState(firstTimeState)
```
maybe something like
`versionedData = localData || diskStore.getState() || migrator.generateInitialState(firstTimeState)`

danfinlay replied:
New commit moves storage.local loading before this assignment so I can use a similar format to what you suggest.

danfinlay committed:
Add clarifying comment

danfinlay committed:
Tighten up data loading code

kumavis reviewed (dismissed):
im going to take another pass over this, please wait

kumavis committed:
background - storage - cleanup storage wiring

kumavis committed:
gitignore - add yarn.lock

kumavis committed:
Merge branch 'master' of github.com:MetaMask/metamask-extension into i3076-UseStorageLocalInstead

kumavis committed:
lint fix

kumavis committed:
Merge branch 'master' of github.com:MetaMask/metamask-extension into i3076-UseStorageLocalInstead

kumavis committed:
Update CHANGELOG.md

kumavis committed:
Merge branch 'master' into i3076-UseStorageLocalInstead

lazaridiscom commented:
@danfinlay , what about the old installations, which use the old storage? Are they migrated to the new storage model?

danfinlay commented:
Yes, they are migrated to the new storage model, and the old storage model is kept around, in case the new one fails.
